neps/nep-0384.md

```diff
@@ -0,0 +1,116 @@
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+
+    provableStore.set(connectionPath(identifier), connection)
+    addConnectionToClient(clientIdentifier, identifier)
+}
+
+// ConnOpenAck relays acceptance of a connection open attempt from chain B back to chain A (this code is executed on chain A).
+function connOpenAck(
+  identifier: Identifier,
+  clientState: ClientState,
+  version: string,
+  counterpartyIdentifier: Identifier,
+  proofTry: CommitmentProof,
+  proofClient: CommitmentProof,
+  proofConsensus: CommitmentProof,
+  proofHeight: Height,
+  consensusHeight: Height) {
+    abortTransactionUnless(consensusHeight < getCurrentHeight())
+    abortTransactionUnless(validateSelfClient(clientState))
+    connection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless((connection.state === INIT && connection.version.indexOf(version) !== -1)
+    expectedConsensusState = getHostConsensusState(consensusHeight) // we need to return a near header at this height.
+    expectedConnectionEnd = ConnectionEnd{TRYOPEN, identifier, getCommitmentPrefix(),
+                             connection.counterpartyClientIdentifier, connection.clientIdentifier,
+                             version, connection.delayPeriodTime, connection.delayPeriodBlocks}
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofTry, counterpartyIdentifier, expectedConnectionEnd)) // so we can verify that the counterparty chain has the correct header.
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, connection.counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+    connection.state = OPEN
+    connection.version = version
+    connection.counterpartyConnectionIdentifier = counterpartyIdentifier
+    provableStore.set(connectionPath(identifier), connection)
+}
+```
+
+The api `env::block_hash` allows us validate the full header submitted via contract calldata by comparing hashes.
+
+## Rationale and alternatives
+
+A few other designs were considered such as a function `env::block_header` which would return the full header at a given height, but was discouraged against a result of expensive db calls. This mapping of block numbers to headers are cheap enough to be stored on-chain.
+
+
+## Specification
+
+```rust
+/// This function should return a blockhash for 256 of the most reccent blocks
+fn block_hash(block_height: u64) -> Option<[u8; 32]>;
+```
+
+This API will require that the map of block heights to hashes are stored in each shard's state trie pending the implementation of a global state trie.
```

Comment:
Just to clarify: The status on global state is not even close to committed, at this point it is just a high-level idea of potential work for the mid to far future. So we really need a solid design for duplicating the state on all shards, with specified binary layout etc.

Reply:
Will this design be part of the NEP?

Reply:
The design should be described in the NEP with reasonable details.

I would expect to see at least a definition of the key-value pairs involved in the trie and what encoding shall be used. Other details, such as padding, can probably be discussed later or left as implementation detail entirely.

Also, the NEP should describe how missing chunks will be handled. If a chunk is skipped, that shard will not have an updated state trie, so it will not be possible to have the latest block hashes. The NEP should define what the block hashes will be available during and after such a scenario.

Reply:
I've started drafting these points you've mentioned. I am clueless on how to deal with the skipped chunks. Can you help me out there?

Reply:
@jakmeier

Reply:
Thanks for the ping.

Yeah, it's a bit tricky. The basic problem is that a shard does not have an updated state root if no new chunk has been produced. Getting the latest block hash is impossible if we want to retrieve it from the state root.

However, do we really need the latest block hash? Maybe a solution would be to define, in the specification, that the host function will always return the block hash of the latest block in which the shard's chunk is present.

My reasoning for why it would be okay: If shard 1 skipped a few chunks, then a contract in shard 1 should not really care about state changes in these blocks. All observable state for that shard is still the same as it was before the skipped chunks.

Counter point: When building bridges, there is an off-chain party calling into the contract that is on an arbitrary shard. This party cares about the global state, not about the arbitrary shard's view of the state.

I don't know enough about the use cases to judge if the counter point is a valid concern. I suspect it might. But I don't really see a solution here otherwise. We would probably have to give up on the idea of storing the block hashes in the state trie and search for more complex alternatives. Such as using some form of global state, as we have also discussed before, I think. (Of course, this would hugely increase the scope.)

Reply:
IBC works by having a light client of the counterparty chain running on NEAR (in this case). This means that we should be able to create this light client once (initialize it), and then maintain its state periodically, to ensure that it's a reasonable up to date representation of what's happening on the other chain. This Light Client, needs to be able to talk to a NEAR light client living on the counterparty chain.

There could be scenario in which the light client in NEAR receives a packet from the other chain, in which it needs to verify a certain state for a block number for which there's no reference on that specific shard. In that case, the request will fail. Even when the execution failed, that shard would get its local state updated for that specific block height, but unfortunately it won't match the block number from the packet (it will be higher).

One potential option could be to retry until the execution of the contract is mapped with the shard that actually contains the right state. However, that's not guaranteed to happen within a certain time frame, and I believe it's not even guaranteed to happen at all, right?

Based on this, seems like we need one design in which either:

- all shards have  access to a global state
- each shard gets its state updated on each block regardless if there was or not a contract being executed
on that specific block (by executing a dummy contract on each shard on every block, for instance)

Reply:
Thanks for the explanation! 

I think there is a misunderstanding for why empty chunks occur. It has nothing to do with whether a contract was executed or not. It's actually because sometimes chunk-producers may miss their chunk, for various reasons. We don't stall the entire chain in that case, instead we let other shards move on and submit their chunk anyway. In which case there will be a block with a missing chunk.

> One potential option could be to retry until the execution of the contract is mapped with the shard that actually contains the right state. However, that's not guaranteed to happen within a certain time frame, and I believe it's not even guaranteed to happen at all, right?

We have a guarantee that a chunk will **eventually** be produced, as long as 2/3 of chunk and block producer work properly. Otherwise the liveness property of the protocol would be violated.
For the exact timing, I'm not sure if we can quantify it. But here we are basically worrying about the case where one shard makes no progress for a long time. No matter how this NEP is implemented, the shard making progress is a pre-condition for the bridge to work anyway. Therefore, I would say that problem is out of scope for this NEP. 

So I think retrying would work and wouldn't slow things down more than fundamentally possible. It doesn't sound super super pretty but maybe it's the best solution, as otherwise, it seems we would require some form of global state to be implemented first. (Which would almost certainly also only see an update on shard x after shard x has produced a chunk.)

Reply:
I'm happy to retry :). I'll update the NEP accordingly then